aptos-move/framework/aptos-framework/sources/jwks.move

```diff
@@ -0,0 +1,706 @@
+            option::some(old_provider_info.config_url)
+        } else {
+            option::none()
+        };
+
+        old_config_endpoint
+    }
+
+    /// Set the `Patches`. Only called in governance proposals.
+    public fun set_patches(aptos_framework: &signer, patches: vector<Patch>) acquires Patches, PatchedJWKs, ObservedJWKs {
+        system_addresses::assert_aptos_framework(aptos_framework);
+        borrow_global_mut<Patches>(@aptos_framework).patches = patches;
+        regenerate_patched_jwks();
+    }
+
+    /// Create a `Patch` that removes all entries.
+    public fun new_patch_remove_all(): Patch {
+        Patch {
+            variant: pack(PatchRemoveAll {}),
+        }
+    }
+
+    /// Create a `Patch` that removes the entry of a given issuer, if exists.
+    public fun new_patch_remove_issuer(issuer: vector<u8>): Patch {
+        Patch {
+            variant: pack(PatchRemoveIssuer { issuer }),
+        }
+    }
+
+    /// Create a `Patch` that removes the entry of a given issuer, if exists.
+    public fun new_patch_remove_jwk(issuer: vector<u8>, jwk_id: vector<u8>): Patch {
+        Patch {
+            variant: pack(PatchRemoveJWK { issuer, jwk_id })
+        }
+    }
+
+    /// Create a `Patch` that upserts a JWK into an issuer's JWK set.
+    public fun new_patch_upsert_jwk(issuer: vector<u8>, jwk: JWK): Patch {
+        Patch {
+            variant: pack(PatchUpsertJWK { issuer, jwk })
+        }
+    }
+
+    /// Create a `JWK` of variant `RSA_JWK`.
+    public fun new_rsa_jwk(kid: String, alg: String, e: String, n: String): JWK {
+        JWK {
+            variant: copyable_any::pack(RSA_JWK {
+                kid,
+                kty: utf8(b"RSA"),
+                e,
+                n,
+                alg,
+            }),
+        }
+    }
+
+    /// Create a `JWK` of variant `UnsupportedJWK`.
+    public fun new_unsupported_jwk(id: vector<u8>, payload: vector<u8>): JWK {
+        JWK {
+            variant: copyable_any::pack(UnsupportedJWK { id, payload })
+        }
+    }
+
+    /// Initialize some JWK resources. Should only be invoked by genesis.
+    public(friend) fun initialize(account: &signer) {
+        system_addresses::assert_aptos_framework(account);
+        move_to(account, SupportedOIDCProviders { providers: vector[] });
+        move_to(account, ObservedJWKs { jwks: AllProvidersJWKs { entries: vector[] } });
+        move_to(account, Patches { patches: vector[] });
+        move_to(account, PatchedJWKs { jwks: AllProvidersJWKs { entries: vector [] } });
+    }
+
+    /// Only used by validators to publish their observed JWK update.
+    ///
+    /// NOTE: It is assumed verification has been done to ensure each update is quorum-certified,
+    /// and its `version` equals to the on-chain version + 1.
+    fun upsert_into_observed_jwks(aptos_framework: &signer, provider_jwks_vec: vector<ProviderJWKs>) acquires ObservedJWKs, PatchedJWKs, Patches {
```

Comment:
Why is this not public?

Reply:
it doesn't have to - it's supposed to be invoked only by vm

Reply:
actually it's a bit better if it's public - it allows governance proposals to modify the observed JWKs directly, which may be useful in some corner cases

Reply:
Ah, so then can I make this one public too, just for consistency & in case governance ever needs to call it?

Reply:
sure